src/deps/uws.zig

```diff
@@ -69,21 +73,455 @@ pub const InternalLoopData = extern struct {
+        }
+    }
+
+    pub fn shutdownRead(this: *UpgradedDuplex) void {
+        if (this.wrapper) |*wrapper| {
+            _ = wrapper.shutdownRead();
+        }
+    }
+
+    pub fn isShutdown(this: *UpgradedDuplex) bool {
+        if (this.wrapper) |wrapper| {
+            return wrapper.isShutdown();
+        }
+        return true;
+    }
+
+    pub fn isClosed(this: *UpgradedDuplex) bool {
+        if (this.wrapper) |wrapper| {
+            return wrapper.isClosed();
+        }
+        return true;
+    }
+
+    pub fn isEstablished(this: *UpgradedDuplex) bool {
+        return !this.isClosed();
+    }
+
+    pub fn ssl(this: *UpgradedDuplex) ?*BoringSSL.SSL {
+        if (this.wrapper) |wrapper| {
+            return wrapper.ssl;
+        }
+        return null;
+    }
+
+    pub fn sslError(this: *UpgradedDuplex) us_bun_verify_error_t {
+        return .{
+            .error_no = this.ssl_error.error_no,
+            .code = @ptrCast(this.ssl_error.code.ptr),
+            .reason = @ptrCast(this.ssl_error.reason.ptr),
+        };
+    }
+
+    pub fn resetTimeout(this: *UpgradedDuplex) void {
+        this.setTimeoutInMilliseconds(this.current_timeout);
+    }
+    pub fn setTimeoutInMilliseconds(this: *UpgradedDuplex, ms: c_uint) void {
+        if (this.event_loop_timer.state == .ACTIVE) {
+            this.vm.timer.remove(&this.event_loop_timer);
+        }
+        this.current_timeout = ms;
+
+        // if the interval is 0 means that we stop the timer
+        if (ms == 0) {
+            return;
+        }
+
+        // reschedule the timer
+        this.event_loop_timer.next = bun.timespec.msFromNow(ms);
+        this.vm.timer.insert(&this.event_loop_timer);
+    }
+    pub fn setTimeout(this: *UpgradedDuplex, seconds: c_uint) void {
+        log("setTimeout({d})", .{seconds});
+        this.setTimeoutInMilliseconds(seconds * 1000);
+    }
+
+    pub fn deinit(this: *UpgradedDuplex) void {
+        log("deinit", .{});
+        // clear the timer
+        this.setTimeout(0);
+
+        if (this.wrapper) |*wrapper| {
+            wrapper.deinit();
+            this.wrapper = null;
+        }
+
+        this.origin.deinit();
+        if (this.onDataCallback.get()) |callback| {
+            JSC.setFunctionData(callback, null);
+            this.onDataCallback.deinit();
+        }
+        if (this.onEndCallback.get()) |callback| {
+            JSC.setFunctionData(callback, null);
+            this.onEndCallback.deinit();
+        }
+        if (this.onWritableCallback.get()) |callback| {
+            JSC.setFunctionData(callback, null);
+            this.onWritableCallback.deinit();
+        }
+        if (this.onCloseCallback.get()) |callback| {
+            JSC.setFunctionData(callback, null);
+            this.onCloseCallback.deinit();
+        }
+        const ssl_error = this.ssl_error;
+        this.ssl_error = .{};
+        if (ssl_error.reason.len > 0) {
+            bun.default_allocator.free(ssl_error.reason);
```

Comment:
can we move this code to a .deinit function on the ssl_error struct?

Reply:
done